File: backend/core/features/summary/summary_service.py
#
#   Imports
#

import uuid

# Perso

from core.shared.enums.trans_type import TransType
from core.features.summary.summary_port import SummaryDBPort

#
#   Services
#

class SummaryService:
    """
        Service for summary operations.
    """
    def __init__(self, summary_db_port: SummaryDBPort):
        self._summary_db_port = summary_db_port

    async def get_balance(
        self,
        user_id: uuid.UUID,
        year: int,
        trans_types: list[TransType]
    ) -> float:
        return await self._summary_db_port.get_balance(user_id, year, trans_types)

File: frontend/src/app/summary/summary.service.ts
import { inject, Injectable } from '@angular/core';
import { AsyncHttpClient } from '../shared/services/async-http-client';
import { BehaviorSubject, Observable, switchMap } from 'rxjs';
import { OptionsService } from '../shared/services/options';
import { toSignal } from '@angular/core/rxjs-interop';

@Injectable({
  providedIn: 'root',
})
export class SummaryService {
  //
  //   Interfaces
  //
  
  private readonly client = inject(AsyncHttpClient);
  private readonly optionsService = inject(OptionsService);

  //
  //   Methods
  //
  
  /**
   * Gets the user's balance depending onn the year
   * and the transaction type.
   * 
   * @returns The balance of the user.
   */
  public getBalance(): Observable<number> {
    return this.client.get<number>(
      "/summary/balance",
      {
        "year": this.optionsService.year(),
        "trans_types": this.optionsService.multipleTypes(),
      }
    );
  }

  //
  //   Refreshable data
  //
  
  private readonly refreshTrigger = new BehaviorSubject<void>(undefined);

  public refresh(): void {
    this.refreshTrigger.next();
  }

  /**
   * The balance of the user. But can be refreshed
   * ussing the refresh() method of the service on
   * the same instance.
   */
  public balance = toSignal(
    this.refreshTrigger.asObservable().pipe(
      switchMap(() => this.getBalance())
    )
  );
}
